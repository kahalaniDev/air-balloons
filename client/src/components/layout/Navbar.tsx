import React from "react";
import useAppDispatch from "../../hooks/useAppDispatch";
import useAppSelector from "../../hooks/useAppSelector";
import { logout } from "../../features/account/slices/accountSlice";
import { AppBar, Button, Toolbar, Typography, Box } from "@mui/material";
import logo from "../../assets/images/logo.png";

type Props = {};

const Navbar: React.FC<Props> = () => {
  const dispatch = useAppDispatch();
  const { username } = useAppSelector((state) => state.account);

  const handleLogout = () => {
    dispatch(logout());
  };

  return (
    <AppBar color="default" position="static" sx={{ height: "10vh" }}>
      <Toolbar sx={{ display: "flex", justifyContent: "space-between", px: 2 }}>
        <img src={logo} alt="logo" height={100} />
        <Box display="flex">
          <Typography
            color="primary"
            variant="h5"
            sx={{ mr: 4, textTransform: "capitalize", fontWeight: "bold" }}
          >
            {username}
          </Typography>
          <Button color="primary" variant="contained" onClick={handleLogout}>
            Logout
          </Button>
        </Box>
      </Toolbar>
    </AppBar>
  );
};

export default Navbar;
