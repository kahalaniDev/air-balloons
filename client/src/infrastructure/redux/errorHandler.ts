import { ApolloError, isApolloError } from "@apollo/client";
import axios, { AxiosError } from "axios";
import { GraphQLError } from "graphql";
import { IAppError } from "../../models/interfaces";

const GLOBAL_STATUS_CODE = 500;
const INTERNAL_ERR_MSG = "Internal server problem, please try again";
const NEWTWORK_ERR_MSG = "Unable to access the server, please try again";

export const errorHandler = (err: unknown) => {
  if (axios.isAxiosError(err)) return axiosErrorBuilder(err);
  if (err instanceof Error && isApolloError(err))
    return apolloErrorBuilder(err);
  return {
    statusCode: 500,
    message: INTERNAL_ERR_MSG,
  };
};

const axiosErrorBuilder = (axiosError: AxiosError) => {
  console.log(axiosError);
  const statusText = axiosError.response?.statusText
    ? axiosError.response!.statusText
    : "";
  const statusCode = axiosError.response!.status;
  switch (statusText) {
    case "Not Found":
      return { statusCode, message: axiosError.response!.data };
    case "Bad Request":
      return { statusCode, message: axiosError.response!.data };
    case "Conflict":
      return { statusCode, message: axiosError.response!.data };
    case "Unauthorized":
      return { statusCode, message: axiosError.response!.data };
    default:
      return { statusCode, message: INTERNAL_ERR_MSG };
  }
  // return {
  //   statusCode: axiosError.response?.status
  //     ? axiosError.response?.data?.statusCode
  //     : axiosError.code,
  //   message: axiosError.response?.data?.message
  //     ? axiosError.response?.data?.message
  //     : INTERNAL_ERR_MSG,
  // };
};

const apolloErrorBuilder = (apolloError: ApolloError): IAppError => {
  if (apolloError.networkError)
    return { statusCode: GLOBAL_STATUS_CODE, message: NEWTWORK_ERR_MSG };
  if (apolloError.graphQLErrors)
    return graphqlErrorBuilder(apolloError.graphQLErrors[0]);
  return { statusCode: GLOBAL_STATUS_CODE, message: INTERNAL_ERR_MSG };
};

const graphqlErrorBuilder = (graphQLError: GraphQLError): IAppError => {
  const statusCode = graphQLError.extensions.codeNum as number;
  switch (graphQLError.extensions.code) {
    case "UNAUTHENTICATED":
    case "BAD_USER_INPUT":
      return { statusCode, message: graphQLError.message };
    default:
      return { statusCode, message: INTERNAL_ERR_MSG };
  }
};
