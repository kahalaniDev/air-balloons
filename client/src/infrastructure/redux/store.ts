import { configureStore, PreloadedState } from "@reduxjs/toolkit";
import accountReducer, {
  initialState as accountInitialState,
} from "../../features/account/slices/accountSlice";
import balloonsReducer, {
  initialState as balloonInitialState,
} from "../../features/balloons/slices/balloonsSlice";
import { isLoggedIn } from "../../features/account/slices/accountSlice";

type InitialState = {
  account?: typeof accountInitialState;
  balloon?: typeof balloonInitialState;
};

export const createStore = (preloadedState?: PreloadedState<InitialState>) =>
  configureStore({
    preloadedState,
    reducer: { account: accountReducer, balloons: balloonsReducer },
  });

const store = createStore();
store.dispatch(isLoggedIn());

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export default store;
