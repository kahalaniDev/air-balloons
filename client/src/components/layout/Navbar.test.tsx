import Navbar from "./Navbar";
import { Provider as ReduxProvider } from "react-redux";
import { render } from "@testing-library/react";
import { createStore } from "../../infrastructure/redux/store";

describe("Navbar", () => {
  test("check if Navbar match to snapshot", () => {
    expect(
      render(
        <ReduxProvider store={createStore()}>
          <Navbar />
        </ReduxProvider>
      ).asFragment()
    ).toMatchSnapshot();
  });
});
