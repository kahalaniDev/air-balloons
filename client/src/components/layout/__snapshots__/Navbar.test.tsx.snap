// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar check if Navbar match to snapshot 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorDefault MuiAppBar-positionStatic css-8krtud-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1423voi-MuiToolbar-root"
    >
      <img
        alt="logo"
        height="100"
        src="logo.png"
      />
      <div
        class="MuiBox-root css-k008qs"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-1291kst-MuiTypography-root"
        />
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Logout
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
